base/lock.jl

```diff
@@ -290,6 +290,64 @@ macro lock_nofail(l, expr)
     end
 end
 
+"""
+  Lockable(value, lock = ReentrantLock())
+
+Creates a `Lockable` object that wraps `value` and
+associates it with the provided `lock`. This object
+supports [`@lock`](@ref), [`lock`](@ref), [`trylock`](@ref),
+[`unlock`](@ref). To access the value, index the lockable object while
+holding the lock.
+
+!!! compat "Julia 1.11"
+    Requires at least Julia 1.11.
+
+## Example
+
+```jldoctest
+julia> locked_list = Lockable(Int[]);
+
+julia> @lock(locked_list, push!(locked_list[], 1)) # must hold the lock to access the value
+1-element Vector{Int64}:
+ 1
+
+julia> lock(summary, locked_list)
+"1-element Vector{Int64}"
+```
+"""
+struct Lockable{T, L <: AbstractLock}
+    value::T
+    lock::L
+end
+
+Lockable(value) = Lockable(value, ReentrantLock())
```

Comment:
drive-by comment: 
- I don't know the history here
-  ReentrantLock is much safer than a non-reentrantlock for general use
-  ReentrantLock is significantly more expensive than a non-reentrant lock (like a factor 10 worse when it is heavily used)
- most code does not need a ReentrantLock.

Reply:
Might be better to not provide a default here?

Reply:
On the contrary, ReentrantLock is slightly faster than non-reentrant lock

Reply:
If the user knows that the access is fast and not reentrant then they can just specify the SpinLock.

Reply:
They should never do that though, since it is a worse sort of lock

Reply:
> On the contrary, ReentrantLock is slightly faster than non-reentrant lock

That is interesting. I assumed that Julia was similar to C++ where std::mutex is considerably cheaper than std::recursive_mutex.

Reply:
Any reference for that claim? It seems odd that C++ would implement std::recursive_mutex so badly, if they already have a good std::mutex. For contrast, note that Apple implements their standard mutex on top of their recursive_mutex (see https://opensource.apple.com/source/Libc/Libc-262/pthreads/pthread_mutex.c.auto.html), so it would be pretty unexpected for this to be badly very implemented as you claim

Reply:
> Any reference for that claim?

Oh dear, I'm going to need a brown paper bag. My apologies.

I ran some tests (linked here: https://gist.github.com/grandinj/d2ba5525bdccea1b72825b5ff6028648)

And the difference appears to be fairly minor - No real difference on Windows, and std::mutex is only 30% faster on Linux.